File: elohim/elohim-render/src/shim/node_crypto.js
// Node `crypto` builtin shim for the SSR runtime.
//
// The Angular server bundle statically imports Node's `crypto` builtin (bare
// `import ... from "crypto"`). deno_core's module loader cannot resolve a bare
// Node builtin, so every render used to die with a raw TypeError before this
// shim existed. This module is served by NodeShimLoader for the `crypto` /
// `node:crypto` specifiers.
//
// SURFACE DISCIPLINE (the bundle's Node-builtin surface -- extend deliberately,
// never auto-stub): only `createHash` is proven to execute on the landing
// render path (multiformats computes content-address digests via
// `createHash("sha256").update(bytes).digest()`), so only `createHash` is
// implemented for real -- backed by the Rust `op_node_crypto_hash` op (sha2).
// Every other member is a LOUD, NAMED stub: a call throws an error naming the
// member, so an unproven path fails clearly ("crypto.randomBytes is not
// implemented") instead of a cryptic "undefined is not a function". Implement
// more members deliberately, one at a time, as render paths prove they need them.
//
// WARNING: pure ASCII only (this file is embedded via include_str! and served as
// module source; keeping it ASCII matches the rest of src/shim and avoids any
// encoding surprises).

function toBytes(data) {
  if (data == null) return new Uint8Array(0);
  if (typeof data === "string") return new TextEncoder().encode(data);
  if (data instanceof Uint8Array) return data;
  if (data instanceof ArrayBuffer) return new Uint8Array(data);
  if (ArrayBuffer.isView(data)) {
    return new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
  }
  return new TextEncoder().encode(String(data));
}

function bytesToHex(bytes) {
  let out = "";
  for (let i = 0; i < bytes.length; i++) {
    out += bytes[i].toString(16).padStart(2, "0");
  }
  return out;
}

// Node `crypto.createHash(algorithm)` -> incremental Hash. Buffers the update
// chunks in JS and hashes once at digest() time via the Rust op (single-update
// is the only observed usage, so buffering is cheap and correct for multi-update
// too). Returns raw bytes (Uint8Array) for `digest()` with no encoding -- what
// multiformats expects -- or a hex string for `digest("hex")`.
function createHash(algorithm) {
  const algo = String(algorithm).toLowerCase();
  const chunks = [];
  const hash = {
    update(data, _inputEncoding) {
      chunks.push(toBytes(data));
      return hash;
    },
    digest(encoding) {
      let total = 0;
      for (let i = 0; i < chunks.length; i++) total += chunks[i].length;
      const joined = new Uint8Array(total);
      let offset = 0;
      for (let i = 0; i < chunks.length; i++) {
        joined.set(chunks[i], offset);
        offset += chunks[i].length;
      }
      const digest = Deno.core.ops.op_node_crypto_hash(algo, joined);
      if (encoding == null) return digest;
      if (encoding === "hex") return bytesToHex(digest);
      throw new Error(
        "elohim-render crypto shim: createHash().digest('" + encoding +
        "') encoding is not implemented (raw bytes or 'hex' only)."
      );
    },
  };
  return hash;
}

// Build a loud, named stub for a member we have not implemented yet. Accessing
// it is fine (feature-detection stays truthy-free is not a concern for this
// bundle, which only ever CALLS members); calling it throws with the member
// name so the failure points precisely at what to implement next.
function notImplemented(name) {
  return function notImplementedStub() {
    throw new Error(
      "elohim-render crypto shim: node crypto." + name + "() is not implemented. " +
      "The SSR server bundle invoked it during render. Implement it deliberately in " +
      "elohim/elohim-render/src/shim/node_crypto.{js,rs} -- the bundle's Node-builtin " +
      "surface, extend deliberately, never auto-stub."
    );
  };
}

// Named exports. ESM linking requires every `import { X } from "crypto"` in the
// bundle to find a matching named export, so these must EXIST even when the
// render path never reads them:
//  - createPrivateKey / createPublicKey / webcrypto: named-imported by the bundle
//    (proven via static grep) but not on the landing render path -> loud stubs.
//  - createHash: real (see above); also named for `import { createHash }`.
//  - randomBytes / randomUUID: common framework members exported as loud stubs so
//    a `import { randomUUID } from "crypto"` links and fails clearly ON CALL.
const createPrivateKey = notImplemented("createPrivateKey");
const createPublicKey = notImplemented("createPublicKey");
const randomBytes = notImplemented("randomBytes");
const randomUUID = notImplemented("randomUUID");
// bare deno_core exposes no Web Crypto global; expose whatever the isolate has
// (typically undefined) rather than fabricate one. The named export must exist
// for linking regardless of whether the render path reads it.
const webcrypto = (typeof globalThis !== "undefined" ? globalThis.crypto : undefined);

export {
  createHash,
  createPrivateKey,
  createPublicKey,
  randomBytes,
  randomUUID,
  webcrypto,
};

// Default export -- `import crypto from "crypto"` binds this object, and the
// bundle accesses members off it (e.g. `_r.createHash(...)`). Carries the real
// createHash plus loud stubs for every member observed in the bundle's static
// import graph, so a default-binding member call throws a named error instead of
// "undefined is not a function".
const cryptoShim = {
  createHash,
  createPrivateKey,
  createPublicKey,
  randomBytes,
  randomUUID,
  webcrypto,
  randomFillSync: notImplemented("randomFillSync"),
  createHmac: notImplemented("createHmac"),
  createCipheriv: notImplemented("createCipheriv"),
  createDecipheriv: notImplemented("createDecipheriv"),
  createSign: notImplemented("createSign"),
  createVerify: notImplemented("createVerify"),
  createECDH: notImplemented("createECDH"),
  generateKeyPair: notImplemented("generateKeyPair"),
  generateKeyPairSync: notImplemented("generateKeyPairSync"),
  diffieHellman: notImplemented("diffieHellman"),
  publicEncrypt: notImplemented("publicEncrypt"),
  privateDecrypt: notImplemented("privateDecrypt"),
  pbkdf2: notImplemented("pbkdf2"),
  pbkdf2Sync: notImplemented("pbkdf2Sync"),
  sign: notImplemented("sign"),
  verify: notImplemented("verify"),
  getHashes: notImplemented("getHashes"),
  getCiphers: notImplemented("getCiphers"),
};

export default cryptoShim;

File: elohim/elohim-render/src/shim/node_crypto.rs
//! Rust half of the Node `crypto` builtin shim + the WebCrypto global.
//!
//! Two ops live here:
//!
//! - `op_node_crypto_hash` backs the Node `createHash(...).digest()` shim with a
//!   real sha2 implementation. The JS half (`node_crypto.js`) is served as a
//!   module by [`crate::shim::loader::NodeShimLoader`] for the `crypto` /
//!   `node:crypto` specifiers.
//! - `op_elohim_get_random_values` backs the WebCrypto `crypto.getRandomValues`
//!   global with REAL OS entropy (`getrandom`, i.e. the Linux `getrandom(2)`
//!   syscall on this native build). The JS half (`node_webcrypto.js`) is an
//!   INIT-TIME global installer (served via `js = [dir ...]`, like
//!   [`crate::shim::node_globals`] / [`crate::shim::node_buffer`]) that installs
//!   `globalThis.crypto` BEFORE any bundle module evaluates -- a jsbn/brorand-style
//!   CSPRNG probe in the libp2p/multiformats chunk feature-detects
//!   `globalThis.crypto.getRandomValues` at module-eval time and throws
//!   "No secure random number generator found" when it is absent.
//!
//! # One source of truth for `crypto`
//!
//! `globalThis.crypto` (installed by `node_webcrypto.js`) and the `node:crypto`
//! module's `webcrypto` export are the SAME object: `node_crypto.js` binds
//! `webcrypto = globalThis.crypto`. So there is exactly one WebCrypto object across
//! the global surface and the module surface, and `crypto.subtle` (a loud, named
//! stub for now -- see below) is likewise single-sourced.
//!
//! # Entropy honesty (non-negotiable for a crypto path)
//!
//! `getRandomValues` MUST be real entropy or absent -- NEVER a PRNG masquerade.
//! This op backs libp2p / multiformats key + nonce generation on the render path,
//! where a predictable stream would be a security defect, not a rendering quirk.
//! `getrandom::fill` reads the OS CSPRNG and errors (rather than degrading) if the
//! source is unavailable.
//!
//! # Surface discipline
//!
//! Intentionally minimal, mirroring [`crate::shim::node_builtins`]: `createHash`
//! (sha256 / sha512) and `getRandomValues` / `randomUUID` are the only crypto
//! primitives proven to execute on the SSR landing render path. `crypto.subtle.*`
//! is a set of LOUD, NAMED stubs -- a render-path call throws a message naming the
//! operation, so an unproven path fails clearly instead of silently. New
//! algorithms / ops are added deliberately as render paths prove they need them --
//! never speculatively stubbed with fake output.

use deno_core::op2;
use deno_error::JsErrorBox;

/// The JS source of the `crypto` shim module, served by `NodeShimLoader` for the
/// `crypto` / `node:crypto` specifiers.
pub(crate) const CRYPTO_SHIM_JS: &str = include_str!("node_crypto.js");

/// Pure hash core: digest `data` with `algorithm`, returning the raw bytes.
///
/// Only the algorithms the bundle actually uses are implemented (sha256,
/// sha512); any other algorithm is a named error rather than a silently
/// wrong-length or fabricated digest. Kept free of deno_core types so it is
/// directly unit-testable without a V8 isolate.
fn hash_bytes(algorithm: &str, data: &[u8]) -> Result<Vec<u8>, String> {
    use sha2::{Digest, Sha256, Sha512};
    match algorithm {
        "sha256" => Ok(Sha256::digest(data).to_vec()),
        "sha512" => Ok(Sha512::digest(data).to_vec()),
        other => Err(format!(
            "elohim-render crypto shim: hash algorithm '{other}' is not implemented \
             (sha256 / sha512 only)"
        )),
    }
}

/// Hash `data` with `algorithm` and return the raw digest bytes.
///
/// Backs the JS shim's `createHash(algorithm).update(data).digest()`.
#[op2]
#[buffer]
fn op_node_crypto_hash(
    #[string] algorithm: &str,
    #[buffer] data: &[u8],
) -> Result<Vec<u8>, JsErrorBox> {
    hash_bytes(algorithm, data).map_err(JsErrorBox::generic)
}

/// Pure entropy core: return `len` bytes from the OS CSPRNG.
///
/// REAL entropy from the OS CSPRNG via `getrandom::fill` -- never a PRNG. On this
/// native build (`RUSTFLAGS=""`) that is the Linux `getrandom(2)` syscall. An
/// unavailable entropy source is a hard error, not a silent degrade to
/// predictable bytes. Enforces the WebCrypto 65536-byte per-call quota. Kept free
/// of deno_core types so it is directly unit-testable without a V8 isolate.
fn random_bytes(len: usize) -> Result<Vec<u8>, String> {
    // Defense in depth: the JS shim already rejects len > 65536 (WebCrypto quota),
    // but never trust the boundary -- an unbounded allocation from a JS-supplied
    // length is the kind of footgun this rung exists to avoid.
    if len > 65536 {
        return Err(
            "elohim-render webcrypto shim: getRandomValues length exceeds the \
             65536-byte quota"
                .to_string(),
        );
    }
    let mut buf = vec![0u8; len];
    getrandom::fill(&mut buf).map_err(|e| {
        format!("elohim-render webcrypto shim: OS entropy source (getrandom) failed: {e}")
    })?;
    Ok(buf)
}

/// Return `len` bytes of cryptographically-secure OS entropy.
///
/// Backs the WebCrypto `crypto.getRandomValues(typedArray)` global (and, through
/// it, `crypto.randomUUID()`). The JS half (`node_webcrypto.js`) copies the
/// returned bytes into the caller's typed array in place and enforces the
/// WebCrypto 65536-byte per-call quota, so `len` here is already bounded; it is
/// re-bounded defensively in [`random_bytes`] regardless. An unavailable entropy
/// source surfaces as a named `JsErrorBox`, never predictable bytes.
#[op2]
#[buffer]
fn op_elohim_get_random_values(#[number] len: usize) -> Result<Vec<u8>, JsErrorBox> {
    random_bytes(len).map_err(JsErrorBox::generic)
}

deno_core::extension!(
    node_crypto_ext,
    ops = [op_node_crypto_hash, op_elohim_get_random_values],
    js = [dir "src/shim", "node_webcrypto.js"],
);

#[cfg(test)]
mod tests {
    use super::*;

    /// sha256("hello") -- the canonical vector the render-level golden test also
    /// asserts, verified here at the hash-core level independent of V8.
    #[test]
    fn hashes_sha256_known_vector() {
        let digest = hash_bytes("sha256", b"hello").expect("sha256 ok");
        let hex: String = digest.iter().map(|b| format!("{b:02x}")).collect();
        assert_eq!(
            hex,
            "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824"
        );
    }

    #[test]
    fn hashes_sha512_len() {
        let digest = hash_bytes("sha512", b"hello").expect("sha512 ok");
        assert_eq!(digest.len(), 64, "sha512 digest is 64 bytes");
    }

    #[test]
    fn unknown_algorithm_errors_named() {
        let err = hash_bytes("md5", b"hello").expect_err("md5 unsupported");
        assert!(err.contains("md5"), "error names the algorithm: {err}");
        assert!(
            err.contains("not implemented"),
            "error is a clear not-implemented message: {err}"
        );
    }

    /// `random_bytes` returns exactly the requested length (WebCrypto fills the
    /// whole typed array), including the zero-length edge.
    #[test]
    fn random_bytes_preserves_length() {
        for len in [0usize, 1, 4, 16, 32, 256, 65536] {
            let bytes = random_bytes(len).expect("entropy ok");
            assert_eq!(bytes.len(), len, "returns exactly {len} bytes");
        }
    }

    /// Two independent 32-byte draws must differ. A constant/zero-filled result
    /// (the failure mode where the op silently returns un-randomized bytes) would
    /// make these equal. Statistical smoke: the odds of two real 32-byte CSPRNG
    /// draws colliding are ~2^-256.
    #[test]
    fn random_bytes_are_non_constant_across_calls() {
        let a = random_bytes(32).expect("entropy ok");
        let b = random_bytes(32).expect("entropy ok");
        assert_ne!(a, b, "two 32-byte CSPRNG draws must differ");
        // And guard the specific all-zero degrade explicitly.
        assert_ne!(
            a,
            vec![0u8; 32],
            "draw must not be all-zero (un-filled) bytes"
        );
    }

    /// A draw beyond the WebCrypto 65536-byte quota is a named error, not an
    /// unbounded allocation.
    #[test]
    fn random_bytes_rejects_over_quota() {
        let err = random_bytes(65537).expect_err("over-quota must error");
        assert!(err.contains("quota"), "error names the quota: {err}");
    }
}
